File: java/src/beecrypt/security/RSAPrivateCrtKeyImpl.java
package beecrypt.security;

import java.io.*;
import java.math.BigInteger;
import java.security.*;
import java.security.interfaces.*;
import java.security.spec.*;

import beecrypt.io.*;

public class RSAPrivateCrtKeyImpl implements RSAPrivateCrtKey, Cloneable {
	private BigInteger _n, _e, _d, _p, _q, _dp, _dq, _qi;
	private byte[] _enc;

	public RSAPrivateCrtKeyImpl(BigInteger modulus, BigInteger publicExponent,
			BigInteger privateExponent, BigInteger primeP, BigInteger primeQ,
			BigInteger primeExponentP, BigInteger primeExponentQ,
			BigInteger crtCoefficient) {
		_n = modulus;
		_e = publicExponent;
		_d = privateExponent;
		_p = primeP;
		_q = primeQ;
		_dp = primeExponentP;
		_dq = primeExponentQ;
		_qi = crtCoefficient;
	}

	public RSAPrivateCrtKeyImpl(RSAPrivateCrtKeySpec spec) {
		_n = spec.getModulus();
		_e = spec.getPublicExponent();
		_d = spec.getPrivateExponent();
		_p = spec.getPrimeP();
		_q = spec.getPrimeQ();
		_dp = spec.getPrimeExponentP();
		_dq = spec.getPrimeExponentQ();
		_qi = spec.getCrtCoefficient();
	}

	public RSAPrivateCrtKeyImpl(RSAPrivateCrtKey key) {
		_n = key.getModulus();
		_e = key.getPublicExponent();
		_d = key.getPrivateExponent();
		_p = key.getPrimeP();
		_q = key.getPrimeQ();
		_dp = key.getPrimeExponentP();
		_dq = key.getPrimeExponentQ();
		_qi = key.getCrtCoefficient();
	}

	public Object clone() {
		return new RSAPrivateCrtKeyImpl(this);
	}

	public boolean equals(Object obj) {
		if (this == obj)
			return true;

		if (obj instanceof RSAPrivateCrtKey) {
			RSAPrivateCrtKey pri = (RSAPrivateCrtKey) obj;

			if (!pri.getModulus().equals(_n))
				return false;
			if (!pri.getPublicExponent().equals(_e))
				return false;
			if (!pri.getPrivateExponent().equals(_d))
				return false;
			if (!pri.getPrimeP().equals(_p))
				return false;
			if (!pri.getPrimeQ().equals(_q))
				return false;
			if (!pri.getPrimeExponentP().equals(_dp))
				return false;
			if (!pri.getPrimeExponentQ().equals(_dq))
				return false;
			if (!pri.getCrtCoefficient().equals(_qi))
				return false;

			return true;
		}
		return false;
	}

	public BigInteger getModulus() {
		return _n;
	}

	public BigInteger getPublicExponent() {
		return _e;
	}

	public BigInteger getPrivateExponent() {
		return _d;
	}

	public BigInteger getPrimeP() {
		return _p;
	}

	public BigInteger getPrimeQ() {
		return _q;
	}

	public BigInteger getPrimeExponentP() {
		return _dp;
	}

	public BigInteger getPrimeExponentQ() {
		return _dq;
	}

	public BigInteger getCrtCoefficient() {
		return _qi;
	}

	public byte[] getEncoded() {
		if (_enc == null) {
			try {
				ByteArrayOutputStream bos = new ByteArrayOutputStream();
				BeeOutputStream bee = new BeeOutputStream(bos);

				bee.writeBigInteger(_n);
				bee.writeBigInteger(_d);
				bee.writeBigInteger(_e);
				bee.writeBigInteger(_p);
				bee.writeBigInteger(_q);
				bee.writeBigInteger(_dp);
				bee.writeBigInteger(_dq);
				bee.writeBigInteger(_qi);
				bee.close();

				_enc = bos.toByteArray();
			} catch (IOException e) {
				// shouldn't occur
			}
		}

		return _enc;
	}

	public String getAlgorithm() {
		return "RSA";
	}

	public String getFormat() {
		return "BEE";
	}
}

File: c++/provider/RSAPrivateCrtKeyImpl.cxx
/*
 * Copyright (c) 2004 Beeyond Software Holding BV
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include "beecrypt/c++/provider/RSAPrivateCrtKeyImpl.h"
#include "beecrypt/c++/io/ByteArrayOutputStream.h"
using beecrypt::io::ByteArrayOutputStream;
#include "beecrypt/c++/beeyond/BeeOutputStream.h"
using beecrypt::beeyond::BeeOutputStream;

using namespace beecrypt::provider;

namespace {
	const String FORMAT_BEE("BEE");
	const String ALGORITHM_RSA("RSA");
}

RSAPrivateCrtKeyImpl::RSAPrivateCrtKeyImpl(const RSAPrivateCrtKey& copy) : _n(copy.getModulus()), _e(copy.getPublicExponent()), _d(copy.getPrivateExponent()), _p(copy.getPrimeP()), _q(copy.getPrimeQ()), _dp(copy.getPrimeExponentP()), _dq(copy.getPrimeExponentQ()), _qi(copy.getCrtCoefficient())
{
	_enc = 0;
}

RSAPrivateCrtKeyImpl::RSAPrivateCrtKeyImpl(const RSAPrivateCrtKeyImpl& copy) : _n(copy._n), _e(copy._e), _d(copy._d), _p(copy._p), _q(copy._q), _dp(copy._dp), _dq(copy._dq), _qi(copy._qi)
{
	_enc = 0;
}

RSAPrivateCrtKeyImpl::RSAPrivateCrtKeyImpl(const BigInteger& n, const BigInteger& e, const BigInteger& d, const BigInteger& p, const BigInteger& q, const BigInteger& dp, const BigInteger& dq, const BigInteger& qi) : _n(n), _e(e), _d(d), _p(p), _q(q), _dp(dp), _dq(dq), _qi(qi)
{
	_enc = 0;
}

RSAPrivateCrtKeyImpl::~RSAPrivateCrtKeyImpl()
{
	delete _enc;
}

bool RSAPrivateCrtKeyImpl::equals(const Object* obj) const throw ()
{
	if (this == obj)
		return true;

	if (obj)
	{
		const RSAPrivateKey* pri = dynamic_cast<const RSAPrivateKey*>(obj);
		if (pri)
		{
			if (pri->getModulus() != _n)
				return false;

			if (pri->getPrivateExponent() != _d)
				return false;

			return true;
		}
	}
	return false;
}

RSAPrivateCrtKeyImpl* RSAPrivateCrtKeyImpl::clone() const throw ()
{
	return new RSAPrivateCrtKeyImpl(*this);
}

const BigInteger& RSAPrivateCrtKeyImpl::getModulus() const throw ()
{
	return _n;
}

const BigInteger& RSAPrivateCrtKeyImpl::getPrivateExponent() const throw ()
{
	return _d;
}

const BigInteger& RSAPrivateCrtKeyImpl::getPublicExponent() const throw ()
{
	return _e;
}

const BigInteger& RSAPrivateCrtKeyImpl::getPrimeP() const throw ()
{
	return _p;
}

const BigInteger& RSAPrivateCrtKeyImpl::getPrimeQ() const throw ()
{
	return _q;
}

const BigInteger& RSAPrivateCrtKeyImpl::getPrimeExponentP() const throw ()
{
	return _dp;
}

const BigInteger& RSAPrivateCrtKeyImpl::getPrimeExponentQ() const throw ()
{
	return _dq;
}

const BigInteger& RSAPrivateCrtKeyImpl::getCrtCoefficient() const throw ()
{
	return _qi;
}

const bytearray* RSAPrivateCrtKeyImpl::getEncoded() const throw ()
{
	if (!_enc)
	{
        try
        {
            ByteArrayOutputStream bos;
            BeeOutputStream bee(bos);

			bee.writeBigInteger(_n);
			bee.writeBigInteger(_d);
			bee.writeBigInteger(_e);
			bee.writeBigInteger(_p);
			bee.writeBigInteger(_q);
			bee.writeBigInteger(_dp);
			bee.writeBigInteger(_dq);
			bee.writeBigInteger(_qi);
            bee.close();

            _enc = bos.toByteArray();
        }
        catch (IOException&)
        {
        }
	}

	return _enc;
}

const String& RSAPrivateCrtKeyImpl::getAlgorithm() const throw ()
{
	return ALGORITHM_RSA;
}

const String* RSAPrivateCrtKeyImpl::getFormat() const throw ()
{
	return &FORMAT_BEE;
}
